examples/static_map/count_by_key_example.cu

```diff
@@ -0,0 +1,162 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuco/static_map.cuh>
+
+#include <thrust/device_vector.h>
+#include <thrust/logical.h>
+#include <thrust/transform.h>
+
+#include <cub/block/block_reduce.cuh>
+
+#include <cuda/std/atomic>
+
+#include <cmath>
+#include <cstddef>
+#include <cstdint>
+#include <iostream>
+#include <limits>
+
+/**
+ * @file count_by_key_example.cu
+ * @brief Demonstrates usage of the device side APIs for individual operations like insert/find in
+ * the context of a count-by-key operation, i.e. for a histogram over keys.
+ *
+ * Individual operations like a single insert or find can be performed in device code via the
+ * static_map "device_view" types.
+ *
+ * @note This example is for demonstration purposes only. It is not intended to show the most
+ * performant way to do the example algorithm.
+ *
+ */
+
+/**
+ * @brief Inserts keys and counts how often they occur in the input sequence.
+ *
+ * @tparam BlockSize CUDA block size
+ * @tparam Map Type of the map returned from static_map::get_device_mutable_view
+ * @tparam KeyIter Input iterator whose value_type convertible to Map::key_type
+ * @tparam UniqueIter Output iterator whose value_type is convertible to uint64_t
+ *
+ * @param[in] map_view View of the map into which inserts will be performed
+ * @param[in] key_begin The beginning of the range of keys to insert
+ * @param[in] num_keys The total number of keys and values
+ * @param[out] num_unique_keys The total number of distinct keys inserted
+ */
+template <int64_t BlockSize, typename Map, typename KeyIter, typename UniqueIter>
+__global__ void count_by_key(Map map_view,
+                             KeyIter keys,
+                             uint64_t num_keys,
+                             UniqueIter num_unique_keys)
+{
+  typedef cub::BlockReduce<uint64_t, BlockSize> BlockReduce;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+
+  int64_t const loop_stride = gridDim.x * BlockSize;
+  int64_t idx               = BlockSize * blockIdx.x + threadIdx.x;
+
+  uint64_t thread_unique_keys = 0;
+  while (idx < num_keys) {
+    // insert key into the map with a count of 1
+    auto [slot, is_new_key] = map_view.insert_and_find({keys[idx], 1});
+    if (is_new_key) {
+      // first occurrence of the key
+      thread_unique_keys++;
+    } else {
+      // key is already in the map -> increment count
+      slot->second.fetch_add(1, cuda::memory_order_relaxed);
+    }
+    idx += loop_stride;
+  }
+
+  // compute number of successfully inserted new keys for each block
+  // and atomically add to the grand total
+  uint64_t block_unique_keys = BlockReduce(temp_storage).Sum(thread_unique_keys);
+  if (threadIdx.x == 0) {
+    cuda::atomic_ref<uint64_t, cuda::thread_scope_device> grid_unique_keys(
+      *thrust::raw_pointer_cast(num_unique_keys));
```

Comment:
nit: this is a thrust-specific operation while the input is a generic iterator. Would static cast work?

Reply:
I think it's safe to use `thrust::raw_pointer_cast` here as the [docs](https://thrust.github.io/doc/group__memory__management_ga8855ccf5c15c0529206c3cc9cf348af1.html#ga8855ccf5c15c0529206c3cc9cf348af1) state:

> creates a "raw" pointer from a pointer-like type, simply returning the wrapped pointer, should it exist.

So it should work fine on _any_ pointer-like type, not just with Thrust facilities.